noun/jets.c

```diff
@@ -218,135 +494,340 @@ u3j_find(u3_noun bat)
+  if ( u3_none != hot ) {
+    u3_noun reg, jax, hap, bal;
+    c3_l    jax_l;
+    u3x_qual(hot, &reg, &jax, &hap, &bal);
+    jax_l = (c3_l) jax;
+    if ( u3_none != (*loc = _cj_reg_find(reg, cor)) ) {
+      act = u3nq(jax_l, u3k(hap), u3k(bal), _cj_jit(jax_l, bat));
     }
+    u3z(hot);
   }
-  return 0;
+  return act;
 }
 
-/* _cj_hot_mine(): in hot state, declare a core.  RETAINS.
-*/
-static c3_l
-_cj_hot_mine(u3_noun mop, u3_noun cor)
+/* _cj_spot(): find location of cor. expensive. RETAIN.
+ *             bas is complicated to make it easy to cache bashes.
+ *             you can safely ignore it by passing NULL. Otherwise,
+ *             if it points to u3_none, and no location is found,
+ *             a bash will be PRODUCED there. if it contains a bash,
+ *             that will be used instead of calling _cj_bash.
+ */
+static u3_weak
+_cj_spot(u3_noun cor, u3_weak* bas)
 {
-  u3_noun p_mop, q_mop, r_mop, hr_mop, tr_mop;
+  u3_weak bak = u3_none,
+          bar = u3_none,
+          reg = u3_none,
+          loc = _cj_spot_cold(cor, &bar);
 
-  u3x_trel(mop, &p_mop, &q_mop, &r_mop);
-  u3x_cell(r_mop, &hr_mop, &tr_mop);
-  {
-    c3_l par_l;
+  if ( NULL == bas ) {
+    bas = &bak;
+  }
 
-    //  Calculate parent axis.
-    //
-    if ( c3y == hr_mop ) {
-      u3_noun cax = u3j_find(u3h(u3r_at(q_mop, cor)));
+  if ( u3_none != bar ) {
+    if ( (u3_none == *bas) ) {
+      *bas = u3k(u3h(bar));
+    }
+    reg = u3k(u3t(bar));
+    u3z(bar);
+  }
 
-      par_l = u3h(u3h(cax));
-      u3z(cax);
+  if ( u3_none == loc ) {
+    if ( u3_none == *bas ) {
+      *bas = _cj_bash(u3h(cor));
     }
-    else par_l = 0;
 
-    return _cj_hot_mean(par_l, mop, u3h(cor));
+    if ( !(u3C.wag_w & u3o_hashless) ) {
+      u3_weak act = _cj_spot_hot(cor, *bas, &loc);
+      if ( u3_none != act ) {
+        reg = _cj_gust(reg, _cj_loc_axe(loc), _cj_loc_pel(loc), u3k(loc));
+        u3h_put(u3R->jed.cod_p, u3h(cor), u3nc(u3k(*bas), u3k(reg)));
+        /* caution: could overwrites old value, debug batteries etc.
+        **          old value contains old _cj_jit (from different
+        **          battery). if we change jit to (map battery *),
+        **          will need to merge with that map here.
+        */
+        u3h_put(u3R->jed.war_p, loc, act);
+      }
+    }
   }
+  if ( u3_none != bak ) {
+    u3z(bak);
+  }
+  if ( u3_none != reg ) {
+    u3z(reg);
+  }
+  return loc;
+}
+
+/* _cj_cast(): create a u3j_fink that can be used to efficiently verify
+ *             that another core is located where this one is. RETAIN.
+ */
+static u3p(u3j_fink)
+_cj_cast(u3_noun cor, u3_noun loc)
+{
+  c3_w     i_w = 0;
+  u3_noun  j, par, bat, dyn, pax,
+           rev = u3_nul,
+           pat = u3h(loc);
+  u3j_fink* fin_u;
+
+  while ( c3n == u3h(pat) ) {
+    bat = u3h(cor);
+    dyn = u3t(pat);
+    pax = u3h(dyn);
+    loc = u3t(dyn);
+    pat = u3h(loc);
+    rev = u3nc(u3nc(u3k(bat), u3k(pax)), rev);
+    cor = u3r_at(pax, cor);
```

Comment:
`u3x_at()` here too? If `cor` is `u3_none`, the subsequent iteration will be an assertion failure.

Reply:
Again no. The core has a known shape.